downloadservice/app.py

```diff
@@ -159,10 +183,94 @@ def login(user):
     return render_template('index.html', user=user, token=token)
 
 
-def get_upstream_session():
+def certificate_validity(certificate):
+    try:
+        x509 = OpenSSL.crypto.load_certificate(
+            OpenSSL.crypto.FILETYPE_PEM, certificate)
+        asn1_time = x509.get_notAfter()
+        return datetime.strptime(asn1_time.decode(), '%Y%m%d%H%M%S%fZ')
+    except Exception:
+        raise CertificateError('invalid certificate')
+
+
+@app.route(url_prefix + '/upload-cert', methods=['POST'])
+@authenticated
+def upload_cert(user):
+    filename = user_to_path_fragment(user) + ".crt"
+    certificate_file = app.config['CTADS_CERTIFICATE_DIR'] + filename
+
+    certificate = request.json.get('certificate')
+    try:
+        if certificate and certificate_validity(certificate).date() > \
+                (date.today()+timedelta(days=1)):
+            return 'certificate validity too long (max 1 day)', 400
+    except CertificateError as e:
+        sentry_sdk.capture_exception(e)
+        return 'invalid certificate', 400
+
+    with open(certificate_file, 'w') as f:
+        f.write(certificate)
+
+    return 'Certificate stored', 200
+
+
+@app.route(url_prefix + '/upload-main-cert', methods=['POST'])
+@authenticated
+def upload_main_cert(user):
+    if not isinstance(user, dict) or user['admin'] is not True:
+        return 'access denied', 401
+
+    data = request.json
+    certificate = data.get('certificate', None)
+    cabundle = data.get('cabundle', None)
+
+    if certificate is None and cabundle is None:
+        return 'missing certificate of cabundle', 400
+    try:
+        if certificate and certificate_validity(certificate).date() > \
+                (date.today()+timedelta(days=1)):
+            return 'certificate validity too long (max 1 day)', 400
+    except CertificateError as e:
+        sentry_sdk.capture_exception(e)
+        return 'invalid certificate', 400
+
+    updated = set()
+    if certificate is not None:
+        with open(app.config['CTADS_CLIENTCERT'], 'w') as f:
+            f.write(certificate)
+            updated.add('Certificate')
+    if cabundle is not None:
+        with open(app.config['CTADS_CABUNDLE'], 'w') as f:
+            f.write(cabundle)
+            updated.add('CABundle')
+
+    return ' and '.join(updated) + ' stored', 200
```

Comment:
please add here information on when they will expire. Also really useful to have json here (and, in fact, in the other one too) since this will be used through API. 
It's hardly ever useful to return just a string, it's generally good to have either json or html.